### PR TITLE
Specification of configs for DMC paper

### PR DESCRIPTION
**Pull request by nielsleadholm**
This is an initial PR to make more explicit the configs we need for the DMC paper.

Many of these configs are not specified yet in code - this PR instead presents in the preamble comments of the config file the requirements for the configs. My suggestion is that, following feedback and any updates on the choice of parameters, we merge the PR, and add these actual configs in code as they are developed (including changes to `dmc_pretrain_experiments.py`.

This is also an attempt to move us closer to others being able to replicate the paper results by referring to a single set of configs.

### CODE REVIEW

**Comment by nielsleadholm** [comment]
It occurred to me over the weekend that the overview would make more sense as a Markdown README rather than a long docstring.

I've also removed the `surf` configs that we don't actually need for the paper (any with the surface agent will use `touch`) so as to simplify the list of experiments.

**Review by scottcanoe — APPROVED** [review]
Looks great, thanks for doing this! I can make any other adjustments to experiment configs, including pretraining experiments. I have a lot of updates to those to merge shortly.

File: dmc_configs/dmc_configs/README.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+## Figure 1 & 2: Diagramatic Figures With No Experiments
+
+## Figure 3: Robust Sensorimotor Inference
+
+Consists of 4 experiments:
+- `dist_agent_1lm` (i.e. no noise)
+- `dist_agent_1lm_noise` - Sensor noise
+- `dist_agent_1lm_randrot` - 14 random rotations  
+- `dist_agent_1lm_randrot_noise`
```
Comment (vkakerbeck) [inline dmc_configs/dmc_configs/README.md:15]: I'm sure we talked about this before but just to make sure it is still the best choice: Why are we using the distant agent instead of the surface agent here? I understand why we are using it for the ViT comparison but it seems like here we could use the surface agent.

File: dmc_configs/dmc_configs/README.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+- 77 objects
+- Goal-state-driven/hypothesis-testing policy active
+- A single LM (no voting)
+
+The main output measure is accuracy and rotation error as a function of noise conditions.
+
+## Default Parameters for Figures 4+
+Unless specified otherwise, the following figures/experiments use:
+- 77 objects
+- 14 random rotations
```
Comment (vkakerbeck) [inline dmc_configs/dmc_configs/README.md:27]: Didn't we say we will test on 5 random rotations last time?

File: dmc_configs/dmc_configs/README.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+- `dist_agent_1lm_randrot_noise`
+- `dist_agent_2lm_randrot_noise`
+- `dist_agent_4lm_randrot_noise`
+- `dist_agent_8lm_randrot_noise`
+- `dist_agent_16lm_randrot_noise`
+
+This means performance is evaluated with:
+- 77 objects
+- Goal-state-driven/hypothesis-testing policy active
+- Sensor noise and 14 random rotations
```
Comment (vkakerbeck) [inline dmc_configs/dmc_configs/README.md:44]: Same here, why 14?

File: dmc_configs/dmc_configs/README.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+Consists of 8 experiments:
+
+### Inference (7 experiments):
+- `dist_agent_1lm_randrot_nohyp_x_percent_5p` - 5% threshold
+- `dist_agent_1lm_randrot_nohyp_x_percent_10p` - 10% threshold
+- `dist_agent_1lm_randrot_nohyp_x_percent_20p` - 20% threshold (default for other experiments)
+- `dist_agent_1lm_randrot_nohyp_x_percent_40p` - 40% threshold
+- `dist_agent_1lm_randrot_nohyp_x_percent_80p` - 80% threshold
+- `dist_agent_1lm_randrot_nohyp_x_percent_100p` - 100% threshold
```
Comment (vkakerbeck) [inline dmc_configs/dmc_configs/README.md:119]: I'm not sure this one will work as it will require all hypotheses except for the winning one to have evidence <0. Givent that there are quite a few similar objects in the ycb dataset and the objects often have a lot of symmetry, I don't think we will ever reach that condition. But I guess we can just see in that case if the mlh at the end of max_steps is correct.

File: dmc_configs/dmc_configs/README.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+### Training (1 experiment):
+- `dist_agent_77obj_1rot_trained`*
+
+This means performance is evaluated with:
```
Comment (vkakerbeck) [inline dmc_configs/dmc_configs/README.md:124]: Why would we evaluate performance here? Isn't this just about FLOPs during training?

File: dmc_configs/dmc_configs/README.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+- No hypothesis-testing**
+- No voting
+
+*Single rotation evaluation due to FLOPs counting overhead, so we will extrapolate total FLOPs to 14 rotations based on this/
+or can compare to a ViT trained on 1 rotation.
+
+**Due to ViT model comparison.
+
+
+The main output measure is accuracy and FLOPs as a function of x-percent threshold.
```
Comment (vkakerbeck) [inline dmc_configs/dmc_configs/README.md:137]: I'm not sure I understand why we are looking at accuracy here. Is this just a copy-paste leftover?

File: dmc_configs/dmc_configs/README.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+  - This config should therefore also use the model trained on the 10 morphologically similar objects.
+- Random rotations
+- Sensor noise
+- Hypothesis-testing policy active
+- No voting
+
+The main output measure is a dendrogram showing evidence score clustering for the 10 objects.
+
+**TODO:**
+- Config needs specification, including training config for similar objects in `dmc_pretrain_experiments.py`.
```
Comment (vkakerbeck) [inline dmc_configs/dmc_configs/README.md:176]: Also will need to run this experiment with SELECTIVE logging on so we get the evidence values to analyze. We could also use the standard 77 object pretrained models and then just create the dendrogram for a subset of the objects for clarity.

**Review by vkakerbeck — APPROVED** [review]
Nice, thanks for writing this up! Just a few minor questions. Also, should we add a bit on the "configs" for ViT training and inference as well?

File: dmc_configs/dmc_configs/README.md (hunk 1 above)
Comment (nielsleadholm) [reply dmc_configs/dmc_configs/README.md:15]: That's a good point, we should get slightly better accuracy with the surface agent. @scottcanoe just thinking do you have any concerns with that? If not I'll add that back.

File: dmc_configs/dmc_configs/README.md (hunk 2 above)
Comment (nielsleadholm) [reply dmc_configs/dmc_configs/README.md:27]: Good catch

File: dmc_configs/dmc_configs/README.md (hunk 5 above)
Comment (nielsleadholm) [reply dmc_configs/dmc_configs/README.md:124]: Yeah this was meant to refer to the inference side of things (where we also want to look at accuracy), I've re-ordered it to be clearer.

File: dmc_configs/dmc_configs/README.md (hunk 6 above)
Comment (nielsleadholm) [reply dmc_configs/dmc_configs/README.md:137]: As above.

File: dmc_configs/dmc_configs/README.md (hunk 4 above)
Comment (vkakerbeck) [reply dmc_configs/dmc_configs/README.md:119]: Yeah the entanglement is because currently `evidence_update_threshold` is set to `=x_percent_threshold`. If we want it to always test all hypotheses then we could set `evidence_update_threshold='all'` in the config. Other options include [int, float, 'mean', 'median'] This doesn't directly allow testing different percentage wise update thresholds here but we could either test and compare ['mean', 'median', 'all', 'x_percent_threshold'] or increment an int value here. This gets to the point that we have several levers to improve accuracy at the cost of computation: 
- `x_percent_threshold` - How certain does the model need to be before it reaches a terminal condition
- `evidence_update_threshold` - How many hypotheses do we update at each step
- `min_steps` - how many steps do we need to take at least to make a classification
- `max_steps` - how many steps do we allow the system to take at most
Some more entangled or smaller factors:
- whether we do hypothesis testing policy (can add computation but also reduce it by reducing number of steps
- `max_nneighbors` = how many neighbors do we consider when updating evidence
- `vote_evidence_threshold` - same as `evidence_update_threshold` but for voting
- `path_similarity_threshold` and `pose_similarity_threshold` - how similar do paths and poses need to be to be classified as the same (reaching terminal condition earlier)
- `required_symmetry_evidence` - how early can we detect symmetry
- `tolerances` - this one is a bit complicates since it also adds a tradeoff between false positives and negatives
- how many poses we sample when Pc1=Pc2 in `get_more_directions_in_plane` (currently 8)

So its not like we just have one parameter we can tweak. I guess for this plot it would be simplest to focus on one of them (x_percent_threshold?) or a combination of them (maybe testing x_percent_threshold in the range of [5-30] and then adding one point where we test all hypotheses in addition?).

File: dmc_configs/dmc_configs/README.md [hunk 8]
```diff
@@ -118,27 +118,27 @@ Consists of 8 experiments:
...
+The main output measure is accuracy and FLOPs as a function of x-percent threshold.
 
-**Due to ViT model comparison.
+### Training (1 experiment):
+- `dist_agent_77obj_1rot_trained`**
 
+**Single rotation evaluation due to FLOPs counting overhead, so we will extrapolate total FLOPs to 14 rotations based on this/or can compare to a ViT trained on 1 rotation.
 
-The main output measure is accuracy and FLOPs as a function of x-percent threshold.
+The main output measure is as a function of whether the ViT or Monty is training.
```
Comment (vkakerbeck) [inline dmc_configs/dmc_configs/README.md:137]: I'm not sure I understand it. Wouldn't it be FLOPs?

**Review by vkakerbeck — APPROVED** [review]
Nice, thanks for the updates!

File: dmc_configs/dmc_configs/README.md (hunk 4 above)
Comment (nielsleadholm) [reply dmc_configs/dmc_configs/README.md:119]: Nice thanks for the detailed summary, yeah I think "x_percent_threshold in the range of [5-30] and then adding one point where we test all hypotheses in addition" is a good suggestion that tests it under a reasonable range of values. The curve doesn't need to be based on evenly spaced points (in fact, the effective FLOPs will probably have a non-linear relationship with these parameters). So to summarize, it would be:

```python
### Inference (7 experiments):
- `dist_agent_1lm_randrot_nohyp_x_percent_5p` - 5% threshold
  - `evidence_update_threshold=x_percent_threshold`  # Evidence update threshold is the same as x-percent threshold for determining convergence.
- `dist_agent_1lm_randrot_nohyp_x_percent_10p` - 10% threshold
- `dist_agent_1lm_randrot_nohyp_x_percent_15p` - 15% threshold
- `dist_agent_1lm_randrot_nohyp_x_percent_20p` - 20% threshold (default for other experiments)
- `dist_agent_1lm_randrot_nohyp_x_percent_30p` - 30% threshold
- `dist_agent_1lm_randrot_nohyp_x_percent_20p_evidence_update_all`
  - 20% x-percent threshold for *convergence*
  - Effectively 100% threshold for which hypotheses to test, i.e. all hypotheses are tested; determined by setting `evidence_update_threshold='all'
  
**Notes:**
- For the first 5 experiments above, x-percent threshold determines the threshold at which the LM determines it has converged. In addition we set `evidence_update_threshold=x_percent_threshold`, so that this same threshold determines which evidence values are updated. 
- For the final experiment, these are separated out as noted.
```

@hlee9212 and @scottcanoe tagging you guys so you are aware of this subtlety in Monty.

File: dmc_configs/dmc_configs/README.md (hunk 8 above)
Comment (nielsleadholm) [reply dmc_configs/dmc_configs/README.md]: Sorry, typo on my part.

File: dmc_configs/dmc_configs/README.md (hunk 4 above)
Comment (vkakerbeck) [reply dmc_configs/dmc_configs/README.md]: That sounds good :) Just a minor remark for when you write this up: Setting `evidence_update_threshold='all' is not equivalent to `evidence_update_threshold=x_percent_threshold` with `x_percent_threshold=100`. That would be `evidence_update_threshold=0`. If we do 'all' we also update hypotheses that have accumulated negative evidence.

File: dmc_configs/dmc_configs/README.md (hunk 1 above)
Comment (nielsleadholm) [reply dmc_configs/dmc_configs/README.md:15]: I discussed this with Scott and I think it's worth leaving this for the moment (i.e. using distant agents) so that experiments are consistent across the core figures (3, 4, 6, 7) in terms of what agent we are using. 

That said, we decided we can run the surface agent as well for these configs so that we at least have the results. I was thinking (can discuss more on Monday) we can maybe do a distant-vs-surface comparison in Figure 4 in terms of the impact of model-free (and not just model-based) policy.
